module/os/linux/zfs/zpl_super.c

```diff
@@ -276,11 +277,17 @@ zpl_test_super(struct super_block *s, void *data)
 {
 	zfsvfs_t *zfsvfs = s->s_fs_info;
 	objset_t *os = data;
+	int match;
 
-	if (zfsvfs == NULL)
+	if (zfsvfs == NULL || os != zfsvfs->z_os)
```

Comment:
Why is this is okay to do outside of the lock if we know that testing `(os == zfsvfs->z_os)` without locking was giving us the wrong results previously?

I suggest that we drop ` || os != zfsvfs->z_os`.

Reply:
Since the os is held in zpl_mount_impl() for the duration of the sget, if it doesn't match, we do not have to worry about that fact changing. The problem only occurs when a z_os is stale and matches the os we are searching for.

Reply:
Is there no way that `zfsvfs->z_os` being stale could cause it to not match `os` when it should match? That is my concern here.

Reply:
My concern was that always taking the lock if that it might impact the performance of a lot of mounts since it would take the lock for every sb for every mount. So, if any teardown locks are held, any mount will have to wait until it is released.

For the case you are concerned about, that would be a multimount (we are looking for an existing sb associated with that dataset). Here, taking the teardown lock would not be enough b/c the sequence:

1. zpl_mount_impl()->dmu_object_hold to get the os
2. in a separate process, the rollback happens (i.e., there is a different os)
3. sget()->spl_test_super() will see the new os in z_os whether we take the lock or not

The good news is that the result of not finding the correct superblock is an EBUSY, which seems appropriate. I confirmed this by changing zpl_test_super() to always return 0 (and with a zfs_dbgmsg to confirm I hit it).

Reply:
There is so much more code involved when the mount operation occurs that I would not be concerned about a performance regression from a lock that is not held very long.

That said, after reading your explanation, I think the current patch just needs a comment explaining about EBUSY and it should be fine.

Reply:
I added some comments